### ISSUE TITLE
`expectAsyncConsoleError` not throwing error

### ISSUE THREAD

**Issue by zhouyx** [ampproject/amphtml#28438]
I tried to use the `expectAsyncConsoleError()` method, but it's not failing a unit test as I expected. Am I missing anything?

Here's my fake test
```
describe('test', () => {
  afterEach(() => {
    console.log('done');
  });
  it('my fake test', () => {
    expectAsyncConsoleError('fake message');
  });
});
```

The test passed. Got
![image](https://user-images.githubusercontent.com/5419904/81997455-a2960e00-9604-11ea-830c-bec6de638f5e.png)

When I tried to debug it,  I noticed the test was run twice and failed the second time.
![image](https://user-images.githubusercontent.com/5419904/81997509-c8bbae00-9604-11ea-8ab1-adb3260073f7.png)

**Comment by zhouyx** [comment]
#28439 is the draft PR.

**Comment by zhouyx** [comment]
cc @micajuine-ho

**Comment by rsimha** [comment]
@zhouyx The source of this bug is #20091, but the root cause lies in https://github.com/mochajs/mocha/issues/1635. Here's my understanding of what is going on:

- If a test cannot predict when an asynchronous expected `console.error` will actually happen, they can call `expectAsyncConsoleError()` at the start
- At the end of each test, we do a check in `afterEach()` to see if the error message was indeed encountered
- Originally, we were calling `throw new Error(...)` in `afterEach()` if an expected error wasn't seen
- The problem with throwing is that subsequent tests will not run (https://github.com/mochajs/mocha/issues/1635)
- In #20091, the `throw new Error(...)` was effectively changed to `this.test.error(new Error(...))`
- It turns out that calling `this.test.error(new Error(...))` in `afterEach` doesn't actually fail the test (as you observed), because it fails a test after it has passed! (see https://github.com/mochajs/mocha/pull/3362, which was reverted)

The only option I can think of is to go back to `throw new Error(...)` in `afterEach` (revert #20091), but this will halt all tests when the first violation of `expectAsyncConsoleError` is encountered. Is this acceptable to you?

**Comment by zhouyx** [comment]
Thank you @rsimha for the detailed explanation. Sign... 
I find #20091 to be very useful. Killing karma due to a failing test sounds bad. 

Can we avoid verifying the console error in `afterEach()` block? What about having `expectAsyncConsoleError()` to return a promise and request individual test to wait for that promise to be resolved.

**Comment by rsimha** [comment]
> I find #20091 to be very useful. Killing karma due to a failing test sounds bad.

Totally [agree](https://github.com/ampproject/amphtml/pull/20091#pullrequestreview-188192268)!

> Can we avoid verifying the console error in `afterEach()` block? What about having `expectAsyncConsoleError()` to return a promise and request individual test to wait for that promise to be resolved.

The original goal of making `expectAsyncConsoleError` a single-line call instead of a wrapper like `allowConsoleError` was to make tests easy to write. With your suggestion, will the usage of `expectAsyncConsoleError` now effectively change to a wrapper that waits for the expected error? (All [existing usage](https://github.com/ampproject/amphtml/search?utf8=%E2%9C%93&q=expectAsyncConsoleError) will need to be converted.)

In any case, I have no objections to your suggestion if it improves the correctness of the test framework. So if you have the bandwidth, go for it!

**Comment by zhouyx** [comment]
Yes, all the existing usage will need to be converted.  Some of them may silently fail today, will also need to fix that after the conversion.  
Can be a good fixit week item : )

**Comment by samouri** [comment]
I also ran into this issue today. Tracked it down to a bug in `karma-esbuild`: https://github.com/karma-runner/karma-mocha/issues/236.
